api/src/lib/ai/ollama.ts

```diff
@@ -0,0 +1,173 @@
+import chalk from "chalk";
+import OpenAI from "openai";
+import logger from "../../logger.js";
+import { isJson } from "../utils.js";
+import { getSystemPrompt, invokeRequestGenerationPrompt } from "./prompts.js";
+import { makeRequestTool } from "./tools.js";
+
+type GenerateRequestOptions = {
+  apiKey?: string;
+  baseUrl?: string;
+  model: string;
+  persona: string;
+  method: string;
+  path: string;
+  handler: string;
+  handlerContext?: string;
+  history?: Array<string>;
+  openApiSpec?: string;
+  middleware?: {
+    handler: string;
+    method: string;
+    path: string;
+  }[];
+  middlewareContext?: string;
+};
+
+/**
+ * Generates request data for a route handler
+ * - uses OpenAI's tool-calling feature.
+ * - returns the request data as JSON.
+ *
+ * See the JSON Schema definition for the request data in the `make_request` tool.
+ */
+export async function generateRequestWithOllama({
+  // NOTE - Ollama does not require an API key
+  apiKey = "ollama",
+  baseUrl = "http://localhost:11434",
+  model = "llama3.1",
+  persona,
+  method,
+  path,
+  handler,
+  handlerContext,
+  history,
+  openApiSpec,
+  middleware,
+  middlewareContext,
+}: GenerateRequestOptions & {
+  apiKey?: string;
+  model?: string;
+  baseUrl?: string;
+}) {
+  logger.debug(
+    "Generating request data with Ollama",
+    `model: ${model}`,
+    `baseUrl: ${baseUrl}`,
+    `persona: ${persona}`,
+    `method: ${method}`,
+    `path: ${path}`,
+    `handler: ${handler}`,
+    // `handlerContext: ${handlerContext}`,
+    // `openApiSpec: ${openApiSpec}`,
+    // `middleware: ${middleware}`,
+    // `middlewareContext: ${middlewareContext}`,
+  );
+  // Remove trailing slash from baseUrl
+  let openaiCompatibleBaseUrl = baseUrl.replace(/\/$/, "");
+  openaiCompatibleBaseUrl = baseUrl.endsWith("/v1") ? baseUrl : `${baseUrl}/v1`;
+
+  const openaiClient = new OpenAI({ apiKey, baseURL: openaiCompatibleBaseUrl });
+  const userPrompt = await invokeRequestGenerationPrompt({
+    persona,
+    method,
+    path,
+    handler,
+    handlerContext,
+    history,
+    openApiSpec,
+    middleware,
+    middlewareContext,
+  });
+
+  const response = await openaiClient.chat.completions.create({
+    // NOTE - llama-3.1-8b is not very good at function calling, but the 70b and 405b variants were pre-trained to be good at it.
+    //        Don't expect a lot from this :grimace:
+    model,
+    // NOTE - We can try to restrict the response to be from this single tool call
+    tool_choice: {
+      type: "function",
+      function: { name: makeRequestTool.function.name },
+    },
+    tools: [makeRequestTool],
+    messages: [
+      {
+        role: "system",
+        content: getSystemPrompt(persona, "ollama"),
+      },
+      {
+        role: "user",
+        content: userPrompt,
+      },
+    ],
+    temperature: 0.04,
```

Comment:
I reduced temperature on the model for more predictable outputs, but this does mean it is more likely to keep returning the same request payload for a given route / handler